packages/engine/lib/nano/src/client.rs

```diff
@@ -40,16 +41,30 @@ impl fmt::Debug for WorkerHandle {
 }
 
 impl Worker {
-    fn new(url: &str, request_rx: spmc::Receiver<Request>) -> Result<Self> {
-        let socket = nng::Socket::new(nng::Protocol::Req0)?;
-
-        let builder = nng::DialerBuilder::new(&socket, url)?;
-        builder.set_opt::<ReconnectMaxTime>(Some(RECONNECT_MAX_TIME))?;
-        builder.set_opt::<ReconnectMinTime>(Some(RECONNECT_MIN_TIME))?;
-        let dialer = builder.start(false).map_err(|e| e.1)?;
-
-        let ctx = nng::Context::new(&socket)?;
-        ctx.set_opt::<ResendTime>(Some(RESEND_TIME))?;
+    fn new(url: &str, request_rx: spmc::Receiver<Request>) -> Result<Self, ()> {
+        let socket =
+            nng::Socket::new(nng::Protocol::Req0).wrap_err("Could not create nng socket")?;
+
+        let builder =
+            nng::DialerBuilder::new(&socket, url).wrap_err("Could not create nng dialer")?;
+        builder
+            .set_opt::<ReconnectMaxTime>(Some(RECONNECT_MAX_TIME))
+            .wrap_err_lazy(|| {
```

Comment:
This seems to add a lot of code. What are the underlying error messages when `set_opt` fails? And also how can it fail? 

I'm not sure we **always** need to add a contextual message that just repeats what the code is doing.

Reply:
The other thing I'm thinking is that if we always have backtraces then this information is already captured, the bits we're more interested in is adding pieces of context like URL's, experiment ID's, and generally spans. 

I don't think we need to recreate the whole back-trace by just repeating the name of the methods.

I think the general guidance for when we want to add context is if it's in the middle of a function and we need to describe what's happening at that time (for instance in a for-loop and we could say failed to do X while doing Y)

Reply:
> What are the underlying error messages when set_opt fails?

I tracked down the possible errors to the [C-API interface](https://nng.nanomsg.org/man/v1.3.2/nng_setopt.3.html). These are the possible errors:

- `NNG_ECLOSED`: Parameter s does not refer to an open socket.
- `NNG_EINVAL`: The value being passed is invalid.
- `NNG_ENOTSUP`: The option opt is not supported.
- `NNG_EREADONLY`: The option opt is read-only.
- `NNG_ESTATE`: The socket is in an inappropriate state for setting this option.

I absolutely agree, that I'm currently re-creating the backtrace.

Reply:
This is how the `Dialer` creation could look like with try blocks:
```rust

let try_dialer: Result<nng::Dialer, ()> = try {
    let builder = nng::DialerBuilder::new(&socket, url)?;
    builder.set_opt::<ReconnectMaxTime>(Some(RECONNECT_MAX_TIME))?;
    builder.set_opt::<ReconnectMinTime>(Some(RECONNECT_MIN_TIME))?;
    builder.start(false).map_err(|(_, error)| error)?
};
let dialer = try_dialer.wrap_err("Could not start nng dialer")?;
```